File: BandPassFilter.java
/**
 * Bandpass filter from an asymmetric Mach-Zehnder interferometer.
 *
 * Two identical directional couplers (gap, L_DC) separated by arms with a
 * path-length difference Delta_L. The cross port forms a bandpass centered
 * where the arm phase Delta_phi = 0.
 *
 * Design from a filter spec (waveguide W/H/Delta + DC gap given):
 *   FSR      -> Delta_L = lambda_c^2 / (n_eff * FSR), snapped to an integer
 *               number of wavelengths so a cross peak sits at lambda_c.
 *   lambda_c -> peak location and the wavelength axis.
 *   passband -> coupler angle theta = kappa(lambda_c) * L_DC, from
 *                 P_cross = sin^2(2 theta) * cos^2(Delta_phi/2);
 *               the absolute -3 dB width spans 0..FSR/2 by under-coupling,
 *               trading peak transmission sin^2(2 theta) below FSR/2.
 *
 * Transfer (matches band_pass_filter.py / mzi):
 *   P_bar = a^2 + b^2 - 2ab cos(Delta_phi), a = cos(kL1)cos(kL2),
 *                                           b = sin(kL1)sin(kL2)
 *
 * Compile: javac BandPassFilter.java
 * Run:     java BandPassFilter <W> <H> <delta%> <gap> <lam_c_um> <FSR_nm> <PB_nm> [span_xFSR=3] [N=600]
 * Example: java BandPassFilter 5.0 5.0 0.75 5.0 1.55 1.6 0.6
 */
public class BandPassFilter {

    public static double nSilica(double wl) {
        double B1 = 0.6961663, B2 = 0.4079426, B3 = 0.8974794;
        double C1 = 0.0684043 * 0.0684043;
        double C2 = 0.1162414 * 0.1162414;
        double C3 = 9.896161 * 9.896161;
        double l2 = wl * wl;
        return Math.sqrt(1.0 + B1*l2/(l2-C1) + B2*l2/(l2-C2) + B3*l2/(l2-C3));
    }
    public static double[] slabFundamentalTE(double wl, double d, double nC, double nCl) {
        if (nC <= nCl) return null;
        double k0 = 2 * Math.PI / wl;
        double a = d / 2;
        double V = k0 * a * Math.sqrt(nC*nC - nCl*nCl);
        double upper = Math.min(Math.PI/2 - 1e-9, V - 1e-12);
        if (upper <= 0) return null;
        double lo = 1e-9, hi = upper, flo = f(lo, V), fhi = f(hi, V);
        if (flo*fhi > 0) return null;
        for (int i = 0; i < 100; i++) {
            double mid = 0.5*(lo+hi), fmid = f(mid, V);
            if (flo*fmid <= 0) { hi = mid; fhi = fmid; }
            else { lo = mid; flo = fmid; }
            if (hi - lo < 1e-13) break;
        }
        double u = 0.5*(lo+hi), w = Math.sqrt(Math.max(V*V - u*u, 0));
        return new double[] { u/a, w/a, Math.sqrt(nC*nC - (u/a/k0)*(u/a/k0)) };
    }
    private static double f(double u, double V) {
        return u * Math.tan(u) - Math.sqrt(Math.max(V*V - u*u, 0));
    }

    /** Returns [kappa, n_eff] at wl, or null. */
    public static double[] kappaNeff(double wl, double W, double H, double dl, double gap) {
        double nC = nSilica(wl), nCl = nC * (1 - dl/100);
        double[] v = slabFundamentalTE(wl, H, nC, nCl);
        if (v == null) return null;
        double[] h = slabFundamentalTE(wl, W, v[2], nCl);
        if (h == null) return null;
        double kx = h[0], gx = h[1], nEff = h[2];
        double k0 = 2*Math.PI/wl, beta = nEff*k0, a = W/2, h2 = kx*kx;
        double kappa = 2*h2*gx*Math.exp(-gx*gap) / (beta*(h2+gx*gx)*(1+gx*a));
        return new double[] { kappa, nEff };
    }

    /** Bar-port power of the 2-coupler MZI, or NaN. */
    public static double mziPbar(double wl, double W, double H, double dl,
                                 double gap1, double L1, double gap2, double L2, double dL) {
        double[] kn1 = kappaNeff(wl, W, H, dl, gap1);
        double[] kn2 = kappaNeff(wl, W, H, dl, gap2);
        if (kn1 == null || kn2 == null) return Double.NaN;
        double kL1 = kn1[0] * L1, kL2 = kn2[0] * L2;
        double dphi = (2*Math.PI*kn1[1]/wl) * dL;
        double a_ = Math.cos(kL1) * Math.cos(kL2);
        double b_ = Math.sin(kL1) * Math.sin(kL2);
        return a_*a_ + b_*b_ - 2*a_*b_*Math.cos(dphi);
    }

    /** Returns [kappaC, neC, m, dL, fsrAchUm, theta, Ldc, peakT, r, clamped] or null. */
    public static double[] design(double W, double H, double dl, double gap,
                                  double lamC, double fsrNm, double pbNm) {
        double[] kn = kappaNeff(lamC, W, H, dl, gap);
        if (kn == null) return null;
        double kappaC = kn[0], neC = kn[1];
        double fsrUm = fsrNm / 1000.0, pbUm = pbNm / 1000.0;
        double dLtarget = lamC*lamC / (neC * fsrUm);
        double m = Math.max(1, Math.round(neC * dLtarget / lamC));
        double dL = m * lamC / neC;
        double fsrAchUm = lamC*lamC / (neC * dL);
        double rRaw = pbUm / fsrAchUm;
        double r = Math.min(Math.max(rRaw, 1e-6), 0.5);
        double theta = 0.5 * Math.asin(Math.min(1.0, Math.sqrt(0.5) / Math.cos(Math.PI * r / 2)));
        double Ldc = theta / kappaC;
        double peakT = Math.sin(2*theta) * Math.sin(2*theta);
        return new double[] { kappaC, neC, m, dL, fsrAchUm, theta, Ldc, peakT, r, rRaw > 0.5 ? 1 : 0 };
    }

    public static void main(String[] args) {
        if (args.length < 7) {
            System.out.println("Usage: java BandPassFilter <W> <H> <delta%> <gap>"
                + " <lam_c_um> <FSR_nm> <PB_nm> [span_xFSR=3] [N=600]");
            System.out.println("Example: java BandPassFilter 5.0 5.0 0.75 5.0 1.55 1.6 0.6");
            return;
        }
        double W = Double.parseDouble(args[0]);
        double H = Double.parseDouble(args[1]);
        double dl = Double.parseDouble(args[2]);
        double gap = Double.parseDouble(args[3]);
        double lamC = Double.parseDouble(args[4]);
        double fsrNm = Double.parseDouble(args[5]);
        double pbNm = Double.parseDouble(args[6]);
        double span = args.length >= 8 ? Double.parseDouble(args[7]) : 3.0;
        int N = args.length >= 9 ? Integer.parseInt(args[8]) : 600;

        double[] d = design(W, H, dl, gap, lamC, fsrNm, pbNm);
        if (d == null) {
            System.out.println("# design failed: no guided mode for this waveguide/wavelength");
            return;
        }
        double kappaC = d[0], neC = d[1];
        int m = (int) d[2];
        double dL = d[3], fsrAchUm = d[4], theta = d[5], Ldc = d[6], peakT = d[7];
        boolean clamped = d[9] != 0;

        System.out.printf("# Bandpass (asymmetric MZI)  W=%s H=%s delta=%s%% gap=%s%n", W, H, dl, gap);
        System.out.printf("# spec: lambda_c=%s um  FSR=%s nm  passband=%s nm%n", lamC, fsrNm, pbNm);
        System.out.printf("#   n_eff(lam_c)   = %.6f%n", neC);
        System.out.printf("#   kappa(lam_c)   = %.6e /um%n", kappaC);
        System.out.printf("#   Delta_L        = %.4f um   (m = %d wavelengths)%n", dL, m);
        System.out.printf("#   FSR achieved   = %.4f nm%n", fsrAchUm * 1000.0);
        System.out.printf("#   coupler angle  = %.4f rad (%.2f deg)%n", theta, Math.toDegrees(theta));
        System.out.printf("#   L_DC (x2)      = %.4f um   gap = %s um%n", Ldc, gap);
        System.out.printf("#   peak T (cross) = %.4f%n", peakT);
        if (clamped) {
            System.out.println("#   NOTE: requested passband > FSR/2; clamped to FSR/2 (3 dB couplers)."
                + " Use a cascade/lattice for narrower flat-top passbands.");
        }
        System.out.println("# wavelength_um, P_bar, P_cross");

        double wlA = lamC - 0.5 * span * fsrAchUm;
        double wlB = lamC + 0.5 * span * fsrAchUm;
        for (int i = 0; i < N; i++) {
            double wl = wlA + (wlB - wlA) * i / (N - 1);
            double pb = mziPbar(wl, W, H, dl, gap, Ldc, gap, Ldc, dL);
            if (Double.isNaN(pb)) System.out.printf("%.6f, NaN, NaN%n", wl);
            else System.out.printf("%.6f, %.6f, %.6f%n", wl, pb, Math.max(0.0, 1.0 - pb));
        }
    }
}

File: band_pass_filter.py
"""Bandpass filter from an asymmetric Mach-Zehnder interferometer.

The filter is a single asymmetric MZI: two identical directional couplers
(gap, L_DC) separated by two arms with a path-length difference Delta_L.
The cross port forms a bandpass centered where the arm phase Delta_phi = 0.

Design from a filter spec (waveguide W/H/Delta + DC gap given):
  * FSR   -> Delta_L = lambda_c^2 / (n_eff * FSR), snapped to an integer number
             of wavelengths so a cross-port peak sits exactly at lambda_c.
  * lambda_c (center) -> peak location and the wavelength axis.
  * passband -> coupler angle theta = kappa(lambda_c) * L_DC.
             For one MZI the cross response is
                 P_cross = sin^2(2 theta) * cos^2(Delta_phi / 2),
             so the absolute -3 dB (P > 0.5) bandwidth is tunable from 0 up to
             FSR/2 by under-coupling; the peak transmission sin^2(2 theta) is
             the price paid for a passband narrower than FSR/2.

Transfer function (matches mzi.py / MZI.java):
    P_bar = a^2 + b^2 - 2ab cos(Delta_phi),  a = cos(kL1)cos(kL2),
                                             b = sin(kL1)sin(kL2)
    P_cross = 1 - P_bar,   Delta_phi = (2 pi n_eff / lambda) * Delta_L

Usage:
    python band_pass_filter.py <W> <H> <delta_pct> <gap> <lam_c_um> <FSR_nm> <PB_nm> [span_xFSR=3] [N=600]
Example (1550 nm center, 1.6 nm FSR, 0.6 nm passband):
    python band_pass_filter.py 5.0 5.0 0.75 5.0 1.55 1.6 0.6
"""

import math
import sys


def n_silica(wl):
    B1, B2, B3 = 0.6961663, 0.4079426, 0.8974794
    C1, C2, C3 = 0.0684043**2, 0.1162414**2, 9.896161**2
    l2 = wl * wl
    return math.sqrt(1 + B1*l2/(l2-C1) + B2*l2/(l2-C2) + B3*l2/(l2-C3))


def slab_te_fundamental(wl, d, n_core, n_clad):
    if n_core <= n_clad:
        return None
    k0 = 2 * math.pi / wl
    a = d / 2
    V = k0 * a * math.sqrt(n_core**2 - n_clad**2)
    upper = min(math.pi/2 - 1e-9, V - 1e-12)
    if upper <= 0:
        return None

    def f(u):
        w = math.sqrt(max(V*V - u*u, 0.0))
        return u * math.tan(u) - w

    lo, hi = 1e-9, upper
    flo, fhi = f(lo), f(hi)
    if flo * fhi > 0:
        return None
    for _ in range(100):
        mid = 0.5 * (lo + hi)
        fmid = f(mid)
        if flo * fmid <= 0:
            hi, fhi = mid, fmid
        else:
            lo, flo = mid, fmid
        if hi - lo < 1e-13:
            break
    u = 0.5 * (lo + hi)
    w_v = math.sqrt(max(V*V - u*u, 0.0))
    return u/a, w_v/a, math.sqrt(n_core**2 - (u/a/k0)**2)


def kappa_neff(wl, W, H, delta, gap):
    """Returns (kappa per um, n_eff) at this wavelength, or None."""
    n_core = n_silica(wl)
    n_clad = n_core * (1 - delta / 100)
    v = slab_te_fundamental(wl, H, n_core, n_clad)
    if v is None:
        return None
    h = slab_te_fundamental(wl, W, v[2], n_clad)
    if h is None:
        return None
    kx, gx, n_eff = h
    k0 = 2 * math.pi / wl
    beta = n_eff * k0
    a = W / 2
    h2 = kx * kx
    kappa = 2 * h2 * gx * math.exp(-gx * gap) / (beta * (h2 + gx*gx) * (1 + gx * a))
    return kappa, n_eff


def mzi_power(wl, W, H, delta, gap1, L1, gap2, L2, dL):
    """Bar-port power of the 2-coupler MZI, or None."""
    kn1 = kappa_neff(wl, W, H, delta, gap1)
    kn2 = kappa_neff(wl, W, H, delta, gap2)
    if kn1 is None or kn2 is None:
        return None
    k1, n_eff = kn1
    k2, _ = kn2
    kL1 = k1 * L1
    kL2 = k2 * L2
    dphi = (2 * math.pi * n_eff / wl) * dL
    a_ = math.cos(kL1) * math.cos(kL2)
    b_ = math.sin(kL1) * math.sin(kL2)
    return a_*a_ + b_*b_ - 2*a_*b_*math.cos(dphi)


def design(W, H, delta, gap, lam_c, fsr_nm, pb_nm):
    """Solve for (Delta_L, L_DC) from the filter spec. Returns a dict or None."""
    kn = kappa_neff(lam_c, W, H, delta, gap)
    if kn is None:
        return None
    kappa_c, ne_c = kn
    fsr_um = fsr_nm / 1000.0
    pb_um = pb_nm / 1000.0

    # Delta_L from FSR, snapped to an integer number of wavelengths so that a
    # cross-port transmission peak (Delta_phi = 2*pi*m) sits exactly at lam_c.
    dL_target = lam_c * lam_c / (ne_c * fsr_um)
    m = max(1, round(ne_c * dL_target / lam_c))
    dL = m * lam_c / ne_c
    fsr_ach_um = lam_c * lam_c / (ne_c * dL)

    # Coupler angle from the requested absolute -3 dB passband (<= FSR/2).
    r = min(max(pb_um / fsr_ach_um, 1e-6), 0.5)
    theta = 0.5 * math.asin(min(1.0, math.sqrt(0.5) / math.cos(math.pi * r / 2)))
    l_dc = theta / kappa_c
    peak_t = math.sin(2 * theta) ** 2

    return {
        "kappa_c": kappa_c, "ne_c": ne_c, "m": m, "dL": dL,
        "fsr_ach_nm": fsr_ach_um * 1000.0, "theta": theta, "l_dc": l_dc,
        "peak_t": peak_t, "r": r, "clamped": pb_um / fsr_ach_um > 0.5,
    }


def main():
    if len(sys.argv) < 8:
        print("Usage: python band_pass_filter.py <W> <H> <delta_pct> <gap>"
              " <lam_c_um> <FSR_nm> <PB_nm> [span_xFSR=3] [N=600]")
        print("Example: python band_pass_filter.py 5.0 5.0 0.75 5.0 1.55 1.6 0.6")
        sys.exit(1)
    W = float(sys.argv[1])
    H = float(sys.argv[2])
    dl = float(sys.argv[3])
    gap = float(sys.argv[4])
    lam_c = float(sys.argv[5])
    fsr_nm = float(sys.argv[6])
    pb_nm = float(sys.argv[7])
    span = float(sys.argv[8]) if len(sys.argv) >= 9 else 3.0
    N = int(sys.argv[9]) if len(sys.argv) >= 10 else 600

    d = design(W, H, dl, gap, lam_c, fsr_nm, pb_nm)
    if d is None:
        print("# design failed: no guided mode for this waveguide/wavelength")
        sys.exit(1)

    l_dc = d["l_dc"]
    dL = d["dL"]
    fsr_ach_um = d["fsr_ach_nm"] / 1000.0

    print(f"# Bandpass (asymmetric MZI)  W={W} H={H} delta={dl}% gap={gap}")
    print(f"# spec: lambda_c={lam_c} um  FSR={fsr_nm} nm  passband={pb_nm} nm")
    print(f"#   n_eff(lam_c)   = {d['ne_c']:.6f}")
    print(f"#   kappa(lam_c)   = {d['kappa_c']:.6e} /um")
    print(f"#   Delta_L        = {dL:.4f} um   (m = {d['m']} wavelengths)")
    print(f"#   FSR achieved   = {d['fsr_ach_nm']:.4f} nm")
    print(f"#   coupler angle  = {d['theta']:.4f} rad ({math.degrees(d['theta']):.2f} deg)")
    print(f"#   L_DC (x2)      = {l_dc:.4f} um   gap = {gap} um")
    print(f"#   peak T (cross) = {d['peak_t']:.4f}")
    if d["clamped"]:
        print("#   NOTE: requested passband > FSR/2; clamped to FSR/2 (3 dB couplers)."
              " Use a cascade/lattice for narrower flat-top passbands.")
    print("# wavelength_um, P_bar, P_cross")

    wl_a = lam_c - 0.5 * span * fsr_ach_um
    wl_b = lam_c + 0.5 * span * fsr_ach_um
    for i in range(N):
        wl = wl_a + (wl_b - wl_a) * i / (N - 1)
        pb = mzi_power(wl, W, H, dl, gap, l_dc, gap, l_dc, dL)
        if pb is None:
            print(f"{wl:.6f}, NaN, NaN")
        else:
            print(f"{wl:.6f}, {pb:.6f}, {max(0.0, 1.0 - pb):.6f}")


if __name__ == "__main__":
    main()
